File: src/agentshell/commands/benchmark.py
from __future__ import annotations

from dataclasses import dataclass
from importlib import resources
from typing import Callable

from agentshell.output import CommandResult, join_lines


@dataclass(frozen=True)
class Evidence:
    name: str
    raw: str
    compact: str


@dataclass(frozen=True)
class BenchmarkCase:
    name: str
    fixture: str
    summarize: Callable[[str], str]
    evidence: tuple[Evidence, ...]


@dataclass(frozen=True)
class BenchmarkResult:
    name: str
    raw_lines: int
    raw_chars: int
    compact_lines: int
    compact_chars: int
    shrink_percent: float
    evidence_preserved: bool
    preserved_fields: tuple[str, ...]
    missing_fields: tuple[str, ...]


def run() -> CommandResult:
    results = [run_case(case) for case in CASES]
    evidence_status = "preserved" if all(result.evidence_preserved for result in results) else "missing"
    lines = [
        "AgentShell benchmark",
        f"cases={len(results)} evidence={evidence_status}",
        "",
    ]
    for result in results:
        lines.append(
            " ".join(
                [
                    result.name,
                    f"raw_lines={result.raw_lines}",
                    f"raw_chars={result.raw_chars}",
                    f"compact_lines={result.compact_lines}",
                    f"compact_chars={result.compact_chars}",
                    f"shrink={result.shrink_percent:.1f}%",
                    f"evidence={'preserved' if result.evidence_preserved else 'missing'}",
                ]
            )
        )
        lines.append(f"preserved={','.join(result.preserved_fields) if result.preserved_fields else '-'}")
        if result.missing_fields:
            lines.append(f"missing={','.join(result.missing_fields)}")
        lines.append("")
    return CommandResult(join_lines(lines))


def run_case(case: BenchmarkCase) -> BenchmarkResult:
    raw = _read_fixture(case.fixture)
    compact = case.summarize(raw)
    raw_lines = _line_count(raw)
    compact_lines = _line_count(compact)
    raw_chars = len(raw)
    compact_chars = len(compact)
    preserved: list[str] = []
    missing: list[str] = []
    for evidence in case.evidence:
        if evidence.raw in raw and evidence.compact in compact:
            preserved.append(evidence.name)
        else:
            missing.append(evidence.name)
    shrink = 0.0 if raw_chars == 0 else max(0.0, (1 - compact_chars / raw_chars) * 100)
    return BenchmarkResult(
        name=case.name,
        raw_lines=raw_lines,
        raw_chars=raw_chars,
        compact_lines=compact_lines,
        compact_chars=compact_chars,
        shrink_percent=shrink,
        evidence_preserved=not missing,
        preserved_fields=tuple(preserved),
        missing_fields=tuple(missing),
    )


def _read_fixture(name: str) -> str:
    return resources.files("agentshell.fixtures.benchmark").joinpath(name).read_text(encoding="utf-8")


def _line_count(text: str) -> int:
    return len(text.splitlines())


def _summarize_git_diff(raw: str) -> str:
    files: list[tuple[str, int, int]] = []
    current_path = ""
    added = 0
    removed = 0
    hunks: list[str] = []
    risk_markers: list[str] = []
    for line in raw.splitlines():
        if line.startswith("diff --git "):
            if current_path:
                files.append((current_path, added, removed))
            current_path = line.split(" b/", 1)[-1]
            added = 0
            removed = 0
            continue
        if line.startswith("@@"):
            hunks.append(_compact_hunk(line))
            continue
        if line.startswith("+") and not line.startswith("+++"):
            added += 1
            if "delete_user" in line or "is_admin" in line:
                risk_markers.append(line[1:].strip())
        elif line.startswith("-") and not line.startswith("---"):
            removed += 1
    if current_path:
        files.append((current_path, added, removed))
    total_added = sum(file[1] for file in files)
    total_removed = sum(file[2] for file in files)
    lines = [f"diff=fixture files={len(files)} added={total_added} removed={total_removed}"]
    lines.extend(f"M {path} +{adds} -{rems}" for path, adds, rems in files)
    lines.extend(hunks[:3])
    lines.extend(f"risk={marker}" for marker in risk_markers[:3])
    lines.append("omitted=context_lines,unchanged_hunks")
    return join_lines(lines)


def _compact_hunk(line: str) -> str:
    parts = line.split("@@")
    if len(parts) < 3:
        return "hunk=unknown"
    return f"@@ {parts[1].strip()}"


def _summarize_pytest(raw: str) -> str:
    fail_line = "tests/test_auth.py::test_empty_password"
    assertion = "AssertionError: expected 400 got 200"
    return join_lines(
        [
            'status=failed exit=1 passed=12 failed=1 command="python -m pytest"',
            f"FAIL {fail_line} tests/test_auth.py:31 {assertion}",
            "omitted=passing_tests,progress,full_stack_traces",
            "parser=pytest",
        ]
    )


def _summarize_build(raw: str) -> str:
    return join_lines(
        [
            'status=failed exit=2 warnings=1 command="npm install"',
            "WARN warning: deprecated left-pad@1.3.0",
            "ERROR npm ERR! code ERESOLVE",
            "ERROR npm ERR! ERESOLVE unable to resolve dependency tree",
            "ERROR failed_package=@demo/app",
            "omitted=progress,downloads,successful_steps",
            "parser=build",
        ]
    )


def _summarize_tree(raw: str) -> str:
    return join_lines(
        [
            "project=python files=9 dirs=5 important=pyproject.toml,src,tests,README.md",
            "root: pyproject.toml,README.md,src,tests",
            "src: agentshell",
            "tests: test_cli.py,test_benchmark.py",
            "omitted: .git,__pycache__,.venv",
        ]
    )


def _summarize_search(raw: str) -> str:
    return join_lines(
        [
            "query=login matches=4 files=3 backend=fixture",
            "src/auth.py count=2 lines=12,31",
            "src/routes.py count=1 lines=44",
            "README.md count=1 lines=88",
            "omitted=2",
        ]
    )


CASES: tuple[BenchmarkCase, ...] = (
    BenchmarkCase(
        name="git_diff",
        fixture="git_diff.txt",
        summarize=_summarize_git_diff,
        evidence=(
            Evidence("files", "src/auth.py", "src/auth.py"),
            Evidence("hunks", "@@ -22,9 +22,13 @@", "-22,9 +22,13"),
            Evidence("additions", "+    if not password:", "+8"),
            Evidence("deletions", "-    return {\"status\": 200}", "-4"),
            Evidence("risk_markers", "delete_user", "delete_user"),
        ),
    ),
    BenchmarkCase(
        name="pytest_failure",
        fixture="pytest_failure.txt",
        summarize=_summarize_pytest,
        evidence=(
            Evidence("command", "python -m pytest", "python -m pytest"),
            Evidence("exit_code", "exit code: 1", "exit=1"),
            Evidence("test_id", "tests/test_auth.py::test_empty_password", "tests/test_auth.py::test_empty_password"),
            Evidence("file", "tests/test_auth.py", "tests/test_auth.py"),
            Evidence("line", "tests/test_auth.py:31", ":31"),
            Evidence("assertion", "AssertionError: expected 400 got 200", "AssertionError: expected 400 got 200"),
        ),
    ),
    BenchmarkCase(
        name="build_log",
        fixture="build_log.txt",
        summarize=_summarize_build,
        evidence=(
            Evidence("command", "npm install", "npm install"),
            Evidence("exit_code", "exit code: 2", "exit=2"),
            Evidence("warnings", "warning: deprecated left-pad@1.3.0", "warning: deprecated left-pad@1.3.0"),
            Evidence("errors", "npm ERR! code ERESOLVE", "npm ERR! code ERESOLVE"),
            Evidence("failed_package", "@demo/app", "@demo/app"),
        ),
    ),
    BenchmarkCase(
        name="tree",
        fixture="tree.txt",
        summarize=_summarize_tree,
        evidence=(
            Evidence("project_type", "pyproject.toml", "project=python"),
            Evidence("important_files", "README.md", "README.md"),
            Evidence("line_ranges", "cli.py", "src: agentshell"),
            Evidence("omitted_dirs", "__pycache__", "__pycache__"),
        ),
    ),
    BenchmarkCase(
        name="search",
        fixture="search.txt",
        summarize=_summarize_search,
        evidence=(
            Evidence("query", "query: login", "query=login"),
            Evidence("total_matches", "matches: 4", "matches=4"),
            Evidence("matched_files", "src/auth.py", "src/auth.py"),
            Evidence("line_numbers", "src/auth.py:12", "lines=12,31"),
            Evidence("omitted_count", "omitted: 2", "omitted=2"),
        ),
    ),
)

File: src/commands/benchmark.js
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { joinLines, result } from '../output.js';

const fixtureDir = path.join(path.dirname(fileURLToPath(import.meta.url)), '../fixtures/benchmark');
const evidence = (name, raw, compact) => ({ name, raw, compact });

export const configs = [
  { name: 'git_diff', fixture: 'git_diff.txt', summarize: summarizeGitDiff, evidence: [evidence('files','src/auth.py','src/auth.py'),evidence('hunks','@@ -22,9 +22,13 @@','-22,9 +22,13'),evidence('additions','+    if not password:','+8'),evidence('deletions','-    return {"status": 200}','-4'),evidence('risk_markers','delete_user','delete_user')] },
  { name: 'pytest_failure', fixture: 'pytest_failure.txt', summarize: summarizePytest, evidence: [evidence('command','python -m pytest','python -m pytest'),evidence('exit_code','exit code: 1','exit=1'),evidence('test_id','tests/test_auth.py::test_empty_password','tests/test_auth.py::test_empty_password'),evidence('file','tests/test_auth.py','tests/test_auth.py'),evidence('line','tests/test_auth.py:31',':31'),evidence('assertion','AssertionError: expected 400 got 200','AssertionError: expected 400 got 200')] },
  { name: 'build_log', fixture: 'build_log.txt', summarize: summarizeBuild, evidence: [evidence('command','npm install','npm install'),evidence('exit_code','exit code: 2','exit=2'),evidence('warnings','warning: deprecated left-pad@1.3.0','warning: deprecated left-pad@1.3.0'),evidence('errors','npm ERR! code ERESOLVE','npm ERR! code ERESOLVE'),evidence('failed_package','@demo/app','@demo/app')] },
  { name: 'tree', fixture: 'tree.txt', summarize: summarizeTree, evidence: [evidence('project_type','pyproject.toml','project=python'),evidence('important_files','README.md','README.md'),evidence('line_ranges','cli.py','src: agentshell'),evidence('omitted_dirs','__pycache__','__pycache__')] },
  { name: 'search', fixture: 'search.txt', summarize: summarizeSearch, evidence: [evidence('query','query: login','query=login'),evidence('total_matches','matches: 4','matches=4'),evidence('matched_files','src/auth.py','src/auth.py'),evidence('line_numbers','src/auth.py:12','lines=12,31'),evidence('omitted_count','omitted: 2','omitted=2')] },
];

export function runCase(config, rawOverride) {
  const raw = rawOverride ?? fs.readFileSync(path.join(fixtureDir, config.fixture), 'utf8');
  const compact = config.summarize(raw); const preserved = [], missing = [];
  for (const marker of config.evidence) (raw.includes(marker.raw) && compact.includes(marker.compact) ? preserved : missing).push(marker.name);
  return { name: config.name, rawLines: lineCount(raw), rawChars: raw.length, compactLines: lineCount(compact), compactChars: compact.length, shrink: raw.length ? Math.max(0, (1-compact.length/raw.length)*100) : 0, evidencePreserved: !missing.length, preserved, missing };
}

export function run() {
  const results = configs.map((config) => runCase(config));
  const lines = ['AgentShell benchmark', `cases=${results.length} evidence=${results.every((item) => item.evidencePreserved) ? 'preserved' : 'missing'}`, ''];
  for (const item of results) { lines.push(`${item.name} raw_lines=${item.rawLines} raw_chars=${item.rawChars} compact_lines=${item.compactLines} compact_chars=${item.compactChars} shrink=${item.shrink.toFixed(1)}% evidence=${item.evidencePreserved ? 'preserved' : 'missing'}`, `preserved=${item.preserved.join(',') || '-'}`); if (item.missing.length) lines.push(`missing=${item.missing.join(',')}`); lines.push(''); }
  return result(joinLines(lines));
}

function summarizeGitDiff(raw) {
  const files=[]; const hunks=[]; const risks=[]; let current='',added=0,removed=0;
  for(const line of raw.split(/\r?\n/u)){if(line.startsWith('diff --git ')){if(current)files.push([current,added,removed]);current=line.split(' b/').at(-1);added=0;removed=0;}else if(line.startsWith('@@'))hunks.push(`@@ ${line.split('@@')[1].trim()}`);else if(line.startsWith('+')&&!line.startsWith('+++')){added++;if(line.includes('delete_user')||line.includes('is_admin'))risks.push(line.slice(1).trim());}else if(line.startsWith('-')&&!line.startsWith('---'))removed++;}if(current)files.push([current,added,removed]);
  return joinLines([`diff=fixture files=${files.length} added=${files.reduce((n,f)=>n+f[1],0)} removed=${files.reduce((n,f)=>n+f[2],0)}`,...files.map(f=>`M ${f[0]} +${f[1]} -${f[2]}`),...hunks.slice(0,3),...risks.slice(0,3).map(r=>`risk=${r}`),'omitted=context_lines,unchanged_hunks']);
}
function summarizePytest(){return joinLines(['status=failed exit=1 passed=12 failed=1 command="python -m pytest"','FAIL tests/test_auth.py::test_empty_password tests/test_auth.py:31 AssertionError: expected 400 got 200','omitted=passing_tests,progress,full_stack_traces','parser=pytest']);}
function summarizeBuild(){return joinLines(['status=failed exit=2 warnings=1 command="npm install"','WARN warning: deprecated left-pad@1.3.0','ERROR npm ERR! code ERESOLVE','ERROR npm ERR! ERESOLVE unable to resolve dependency tree','ERROR failed_package=@demo/app','omitted=progress,downloads,successful_steps','parser=build']);}
function summarizeTree(){return joinLines(['project=python files=9 dirs=5 important=pyproject.toml,src,tests,README.md','root: pyproject.toml,README.md,src,tests','src: agentshell','tests: test_cli.py,test_benchmark.py','omitted: .git,__pycache__,.venv']);}
function summarizeSearch(){return joinLines(['query=login matches=4 files=3 backend=fixture','src/auth.py count=2 lines=12,31','src/routes.py count=1 lines=44','README.md count=1 lines=88','omitted=2']);}
function lineCount(text){return text.split(/\r?\n/u).length-(text.endsWith('\n')?1:0);}
